File: android/app/src/main/java/com/reader/app/nostr/StrictJson.kt
package com.reader.app.nostr

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject

/** kotlinx.serialization parsing with duplicate-object-key rejection at every depth. */
object StrictJson {
  fun parse(text: String): JsonElement {
    Guard(text).check()
    return Json.parseToJsonElement(text)
  }

  fun parseObject(text: String): JsonObject = parse(text).jsonObject

  private class Guard(private val text: String) {
    private var index = 0

    fun check() {
      whitespace()
      value()
      whitespace()
      require(index == text.length) { "trailing JSON data" }
    }

    private fun whitespace() {
      while (index < text.length && text[index] in charArrayOf('\t', '\n', '\r', ' ')) index++
    }

    private fun value() {
      whitespace()
      when (text.getOrNull(index)) {
        '{' -> objectValue()
        '[' -> arrayValue()
        '"' -> stringValue()
        't' -> literal("true")
        'f' -> literal("false")
        'n' -> literal("null")
        else -> numberValue()
      }
    }

    private fun objectValue() {
      index++
      whitespace()
      if (text.getOrNull(index) == '}') { index++; return }
      val keys = mutableSetOf<String>()
      while (true) {
        whitespace()
        require(text.getOrNull(index) == '"') { "object key must be a string" }
        val key = stringValue()
        require(keys.add(key)) { "duplicate JSON key: $key" }
        whitespace()
        require(text.getOrNull(index) == ':') { "missing JSON colon" }
        index++
        value()
        whitespace()
        when (text.getOrNull(index++)) {
          '}' -> return
          ',' -> Unit
          else -> throw IllegalArgumentException("invalid JSON object delimiter")
        }
      }
    }

    private fun arrayValue() {
      index++
      whitespace()
      if (text.getOrNull(index) == ']') { index++; return }
      while (true) {
        value()
        whitespace()
        when (text.getOrNull(index++)) {
          ']' -> return
          ',' -> Unit
          else -> throw IllegalArgumentException("invalid JSON array delimiter")
        }
      }
    }

    private fun stringValue(): String {
      require(text.getOrNull(index++) == '"') { "invalid JSON string" }
      val out = StringBuilder()
      while (index < text.length) {
        val char = text[index++]
        if (char == '"') return out.toString()
        require(char.code >= 0x20) { "unescaped JSON control character" }
        if (char != '\\') { out.append(char); continue }
        val escape = text.getOrNull(index++) ?: throw IllegalArgumentException("unterminated JSON escape")
        when (escape) {
          '"', '\\', '/' -> out.append(escape)
          'b' -> out.append('\b')
          'f' -> out.append('\u000c')
          'n' -> out.append('\n')
          'r' -> out.append('\r')
          't' -> out.append('\t')
          'u' -> {
            val end = index + 4
            require(end <= text.length) { "invalid JSON unicode escape" }
            val digits = text.substring(index, end)
            require(digits.all { it in '0'..'9' || it in 'a'..'f' || it in 'A'..'F' }) { "invalid JSON unicode escape" }
            out.append(digits.toInt(16).toChar())
            index = end
          }
          else -> throw IllegalArgumentException("invalid JSON escape")
        }
      }
      throw IllegalArgumentException("unterminated JSON string")
    }

    private fun literal(expected: String) {
      require(text.regionMatches(index, expected, 0, expected.length)) { "invalid JSON literal" }
      index += expected.length
    }

    private fun digit(char: Char?): Boolean = char != null && char in '0'..'9'

    private fun numberValue() {
      val start = index
      if (text.getOrNull(index) == '-') index++
      if (text.getOrNull(index) == '0') {
        index++
        require(!digit(text.getOrNull(index))) { "invalid leading zero" }
      } else {
        val first = text.getOrNull(index)
        require(first != null && first in '1'..'9') { "invalid JSON value" }
        while (digit(text.getOrNull(index))) index++
      }
      if (text.getOrNull(index) == '.') {
        index++
        require(digit(text.getOrNull(index))) { "invalid JSON fraction" }
        while (digit(text.getOrNull(index))) index++
      }
      if (text.getOrNull(index) == 'e' || text.getOrNull(index) == 'E') {
        index++
        if (text.getOrNull(index) == '+' || text.getOrNull(index) == '-') index++
        require(digit(text.getOrNull(index))) { "invalid JSON exponent" }
        while (digit(text.getOrNull(index))) index++
      }
      require(index > start) { "invalid JSON number" }
    }
  }
}

File: chrome/src/protocol/strict-json.ts
/** JSON.parse plus duplicate-object-key rejection at every nesting level. */
class JsonGuard {
  private index = 0;
  private depth = 0;

  constructor(private readonly text: string) {}

  check(): void {
    this.ws();
    this.value();
    this.ws();
    if (this.index !== this.text.length) throw new Error("trailing JSON data");
  }

  private ws(): void {
    while (this.index < this.text.length && /[\u0009\u000a\u000d\u0020]/.test(this.text[this.index]!)) this.index += 1;
  }

  private value(): void {
    if (++this.depth > 32) throw new Error("JSON nesting limit");
    this.ws();
    const char = this.text[this.index];
    if (char === "{") this.object();
    else if (char === "[") this.array();
    else if (char === '"') void this.string();
    else if (char === "t") this.literal("true");
    else if (char === "f") this.literal("false");
    else if (char === "n") this.literal("null");
    else this.number();
    this.depth--;
  }

  private object(): void {
    this.index += 1;
    this.ws();
    if (this.text[this.index] === "}") { this.index += 1; return; }
    const keys = new Set<string>();
    while (true) {
      this.ws();
      if (this.text[this.index] !== '"') throw new Error("object key must be a string");
      const key = this.string();
      if (keys.has(key)) throw new Error(`duplicate JSON key: ${key}`);
      keys.add(key);
      this.ws();
      if (this.text[this.index] !== ":") throw new Error("missing JSON colon");
      this.index += 1;
      this.value();
      this.ws();
      const delimiter = this.text[this.index++];
      if (delimiter === "}") return;
      if (delimiter !== ",") throw new Error("invalid JSON object delimiter");
    }
  }

  private array(): void {
    this.index += 1;
    this.ws();
    if (this.text[this.index] === "]") { this.index += 1; return; }
    while (true) {
      this.value();
      this.ws();
      const delimiter = this.text[this.index++];
      if (delimiter === "]") return;
      if (delimiter !== ",") throw new Error("invalid JSON array delimiter");
    }
  }

  private string(): string {
    if (this.text[this.index++] !== '"') throw new Error("invalid JSON string");
    let value = "";
    while (this.index < this.text.length) {
      const char = this.text[this.index++]!;
      if (char === '"') return value;
      if (char.charCodeAt(0) < 0x20) throw new Error("unescaped JSON control character");
      if (char !== "\\") { value += char; continue; }
      const escape = this.text[this.index++];
      if (escape === undefined) throw new Error("unterminated JSON escape");
      const simple: Record<string, string> = {
        '"': '"', "\\": "\\", "/": "/", b: "\b", f: "\f", n: "\n", r: "\r", t: "\t",
      };
      if (Object.hasOwn(simple, escape)) {
        value += simple[escape];
      } else if (escape === "u") {
        const digits = this.text.slice(this.index, this.index + 4);
        if (!/^[0-9a-fA-F]{4}$/.test(digits)) throw new Error("invalid JSON unicode escape");
        value += String.fromCharCode(Number.parseInt(digits, 16));
        this.index += 4;
      } else {
        throw new Error("invalid JSON escape");
      }
    }
    throw new Error("unterminated JSON string");
  }

  private literal(expected: string): void {
    if (this.text.slice(this.index, this.index + expected.length) !== expected) throw new Error("invalid JSON literal");
    this.index += expected.length;
  }

  private number(): void {
    const start = this.index;
    if (this.text[this.index] === "-") this.index += 1;
    if (this.text[this.index] === "0") {
      this.index += 1;
      if (/[0-9]/.test(this.text[this.index] ?? "")) throw new Error("invalid leading zero");
    } else {
      if (!/[1-9]/.test(this.text[this.index] ?? "")) throw new Error("invalid JSON value");
      while (/[0-9]/.test(this.text[this.index] ?? "")) this.index += 1;
    }
    if (this.text[this.index] === ".") {
      this.index += 1;
      if (!/[0-9]/.test(this.text[this.index] ?? "")) throw new Error("invalid JSON fraction");
      while (/[0-9]/.test(this.text[this.index] ?? "")) this.index += 1;
    }
    if (this.text[this.index] === "e" || this.text[this.index] === "E") {
      this.index += 1;
      if (this.text[this.index] === "+" || this.text[this.index] === "-") this.index += 1;
      if (!/[0-9]/.test(this.text[this.index] ?? "")) throw new Error("invalid JSON exponent");
      while (/[0-9]/.test(this.text[this.index] ?? "")) this.index += 1;
    }
    if (this.index === start) throw new Error("invalid JSON number");
  }
}

export function parseStrictJson(text: string): unknown {
  new JsonGuard(text).check();
  return JSON.parse(text) as unknown;
}
